### PR TITLE
Imaginary friend QoL (pointing, runechat, hear out of range)

### PR DESCRIPTION
**Pull request by itsmeow**
## About The Pull Request

Imaginary friends now have runechat and their points are visible to the player (and only the player).

If the player says something while the friend can't see them, the message is still sent to the imaginary friend's chat.

If you turn, your icon will update properly and not next time you move.

Unfortunately there is a problem with pointing, where the layer is just above the turf layer, even if I try setting the layer to `POINT_LAYER`. I'm not sure why this is happening. It is functional enough though, you can see the points.

## Why It's Good For The Game

Imaginary friends are really unpolished, this makes them more usable.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

Hearing out of range

image

Runechat (imaginary friend perspective)

image

Runechat (player perspective)
image

Pointing

image

</details>

## Changelog
:cl:
add: Imaginary friends now have runechat.
add: Imaginary friends now have visible points.
fix: Fixed a possible hard-delete with imaginary friend actions.
tweak: Imaginary friends can now hear their owner if out of range.
fix: Imaginary friends now properly turn when they turn instead of next time they move.
/:cl:

### CODE REVIEW

File: code/datums/brain_damage/imaginary_friend.dm [hunk 1]
```diff
@@ -100,6 +100,20 @@
 	hide = new
 	hide.Grant(src)
 
+	// Update icon on turn
+	RegisterSignal(src, COMSIG_ATOM_DIR_CHANGE, .proc/Show)
```
Comment (Archanial) [inline code/datums/brain_damage/imaginary_friend.dm:104]: Does show already use signal handler?

File: code/datums/brain_damage/imaginary_friend.dm [hunk 2]
```diff
@@ -136,6 +150,13 @@
 		client.images.Remove(human_image)
 	return ..()
 
+/mob/camera/imaginary_friend/proc/owner_speech(speaker, speech_args)
+	SIGNAL_HANDLER
+	var/list/listening = get_hearers_in_view(6, owner, SEE_INVISIBLE_MAXIMUM)
+	if(!(src in listening))
+		to_chat(src, "<span class='hear'>You hear a distant voice in your head...</span>")
+		to_chat(src, "<span class='game say'><span class='name'>[speaker]</span> <span class='message'>[say_quote(speech_args[SPEECH_MESSAGE])]</span></span>")
```
Comment (Archanial) [inline code/datums/brain_damage/imaginary_friend.dm:159]: Why are you doubling spans here?
Do classes game say exist?

File: code/datums/brain_damage/imaginary_friend.dm [hunk 3]
```diff
@@ -153,25 +174,32 @@
...
 		return
 
 	src.log_talk(message, LOG_SAY, tag="imaginary friend")
 
+	// Display message
+	if (!owner.client?.prefs.chat_on_map)
+		var/mutable_appearance/MA = mutable_appearance('icons/mob/talk.dmi', src, "default[say_test(message)]", FLY_LAYER)
+		MA.appearance_flags = APPEARANCE_UI_IGNORE_ALPHA
+		INVOKE_ASYNC(GLOBAL_PROC, /proc/flick_overlay, MA, list(owner.client), 30)
+	if(owner.client?.prefs.chat_on_map || client?.prefs.chat_on_map)
```
Comment (Archanial) [inline code/datums/brain_damage/imaginary_friend.dm]: First part is already checked above

Comment (Archanial) [reply code/datums/brain_damage/imaginary_friend.dm]: Actually you are checking it 3 times. Perhaps save result to bool?

File: code/datums/brain_damage/imaginary_friend.dm [hunk 4]
```diff
@@ -153,25 +174,32 @@
...
+		var/list/hearers = list()
+		if(client?.prefs.chat_on_map)
+			hearers += client
+		if(owner.client?.prefs.chat_on_map)
+			hearers += owner.client
+		new /datum/chatmessage(message, src, hearers, null)
```
Comment (Archanial) [inline code/datums/brain_damage/imaginary_friend.dm:198]: Isn't there some wrapper proc you should call before new chatmessages? I'd have to check myself but I vaguely remember it existing

File: code/datums/brain_damage/imaginary_friend.dm (hunk 2 above)
Comment (itsmeow) [reply code/datums/brain_damage/imaginary_friend.dm:159]: This is copied from existing say code, it's for chat tab filters.

File: code/datums/brain_damage/imaginary_friend.dm (hunk 4 above)
Comment (itsmeow) [reply code/datums/brain_damage/imaginary_friend.dm:198]: yes but it basically does a bunch of logic for hearing, language, and emotes that don't work in this case

File: code/datums/brain_damage/imaginary_friend.dm (hunk 1 above)
Comment (itsmeow) [reply code/datums/brain_damage/imaginary_friend.dm:104]: added

File: code/datums/brain_damage/imaginary_friend.dm (hunk 2 above)
Comment (Archanial) [reply code/datums/brain_damage/imaginary_friend.dm:159]: copy or not you can reduce the amount of spans by combining game say with name

Comment (itsmeow) [reply code/datums/brain_damage/imaginary_friend.dm:159]: No, because name only wraps the name, the message won't filter properly otherwise

**Review by PowerfulBacon — APPROVED** [review]

**Pull request merged**